File: EMU7800.D2D/EMU7800.D2D.Interop.WinRT_8.1WP/TextLayout.cpp
// © Mike Murphy

#include "pch.h"
#include "TextLayout.h"

using namespace EMU7800::D2D::Interop;

TextLayout::TextLayout(IDWriteFactory* pDWriteFactory, String^ fontFamilyName, int fontWeight, int fontStyle, int fontStretch, FLOAT fontSize, String ^text, FLOAT width, FLOAT height) :
    m_pTextLayout(0),
    m_hr(0)
{
    if (!pDWriteFactory)
    {
        m_hr = E_POINTER;
        return;
    }

    m_hr = pDWriteFactory->CreateTextFormat(
        fontFamilyName->Data(),
        NULL,
        (DWRITE_FONT_WEIGHT)fontWeight,
        (DWRITE_FONT_STYLE)fontStyle,
        (DWRITE_FONT_STRETCH)fontStretch,
        fontSize,
        L"",
        &m_pTextFormat
        );
    if FAILED(m_hr)
        return;

    m_hr = pDWriteFactory->CreateTextLayout(
        text->Data(),
        text->Length(),
        m_pTextFormat.Get(),
        width,
        height,
        &m_pTextLayout
        );
    if SUCCEEDED(m_hr)
        RefreshMetrics();
}

void TextLayout::RefreshMetrics()
{
    DWRITE_TEXT_METRICS m = {0};
    m_pTextLayout->GetMetrics(&m);
    m_width = m.width;
    m_height = m.height;
    m_lineCount = m.lineCount;
}

int TextLayout::SetTextAlignment(DWriteTextAlignment textAlignment)
{
    if FAILED(m_hr)
        return m_hr;

    HRESULT hr = m_pTextLayout->SetTextAlignment((DWRITE_TEXT_ALIGNMENT)textAlignment);
    if SUCCEEDED(hr)
        RefreshMetrics();
    return hr;
}

int TextLayout::SetParagraphAlignment(DWriteParaAlignment paragraphAlignment)
{
    if FAILED(m_hr)
        return m_hr;

    HRESULT hr = m_pTextLayout->SetParagraphAlignment((DWRITE_PARAGRAPH_ALIGNMENT)paragraphAlignment);
    if SUCCEEDED(hr)
        RefreshMetrics();
    return hr;
}

File: src/win32.interop/TextLayout.cs
using System;

using static EMU7800.Win32.Interop.Direct2DNativeMethods;

namespace EMU7800.Win32.Interop
{
    public class TextLayout : IDisposable
    {
        #region Fields

        readonly string _fontFamilyName, _text;
        readonly int _fontWeight, _fontStyle, _fontStretch;
        readonly float _fontSize, _width, _height;

        public IntPtr TextFormatPtr { get; private set; }
        public IntPtr TextLayoutPtr { get; private set; }

        public int HR { get; private set; }

        #endregion

        public void SetTextAlignment(DWriteTextAlignment textAlignment)
            => Direct2D_SetTextAlignmentForTextLayout(TextLayoutPtr, textAlignment);

        public void SetParagraphAlignment(DWriteParaAlignment paragraphAlignment)
            => Direct2D_SetParagraphAlignmentForTextLayout(TextLayoutPtr, paragraphAlignment);

        public void Draw(D2D_POINT_2F location, D2DSolidColorBrush brush)
            => Direct2D_DrawTextLayout(TextLayoutPtr, location, brush);

        public void GetMetrics(out DWRITE_TEXT_METRICS metrics)
            => Direct2D_GetMetrics(TextLayoutPtr, out metrics);

        public void Initialize()
        {
            if (TextLayoutPtr != IntPtr.Zero)
                return;
            var fptr = TextFormatPtr;
            var lptr = TextLayoutPtr;
            HR = Direct2D_CreateTextLayout(_fontFamilyName, _fontWeight, _fontStyle, _fontStretch, _fontSize, _text, _width, _height, ref fptr, ref lptr);
            TextFormatPtr = fptr;
            TextLayoutPtr = lptr;
        }

        #region IDispose Members

        public void Dispose()
        {
            if (TextLayoutPtr == IntPtr.Zero)
                return;
            Direct2D_ReleaseTextLayout(TextFormatPtr, TextLayoutPtr);
            TextFormatPtr = TextLayoutPtr = IntPtr.Zero;
            HR = 0;
            GC.SuppressFinalize(this);
        }

        #endregion

        #region Constructors

        public TextLayout(string fontFamilyName, float fontSize, string text, float width, float height)
            : this(fontFamilyName,
                   DWRITE_FONT_WEIGHT_NORMAL,
                   DWRITE_FONT_STYLE_NORMAL,
                   DWRITE_FONT_STRETCH_NORMAL,
                   fontSize,
                   text,
                   width,
                   height) {}

        public TextLayout(string fontFamilyName, int fontWeight, int fontStyle, int fontStretch, float fontSize, string text, float width, float height)
        {
            _fontFamilyName = fontFamilyName;
            _fontWeight     = fontWeight;
            _fontStyle      = fontStyle;
            _fontStretch    = fontStretch;
            _fontSize       = fontSize;
            _text           = text;
            _width          = width;
            _height         = height;
            Initialize();
        }

        #endregion
    }
}
